[5076] Publication Page Authors

KalobTaulien opened the pull request:
Closes #5076 
Related PRs/issues #5076 

This work fetches the authors from any child or grandchild ArticlePage (which come from Orderables). Should we need to do this a different way we might want to consider closing this PR and opening a new one with similar (but different) work. 

It's the blog authors for the Publication Hero section, like below: 
![image](https://user-images.githubusercontent.com/4743971/91348166-a52a0700-e7a0-11ea-9bad-ffeb76b3fcc3.png)

easherma commented on network-api/networkapi/wagtailpages/templates/wagtailpages/publication_page.html:
```diff
@@ -32,8 +32,29 @@ <h1 class="h1-heading">
 
-                    {% if page.publication_file %}
 
+                    <div class="blog-authors d-flex flex-wrap align-items-center">
```
Is this template copy-pasted from elsewhere or is it different? I'm wondering if we have the opportunity to make a template block here.

easherma commented on network-api/networkapi/wagtailpages/pagemodels/publications/publication.py:
```diff
@@ -80,3 +80,12 @@ class PublicationPage(FoundationMetadataPageMixin, Page):
         FieldPanel('notes')
     ]
+
+    def get_authors(self) -> set:
```
I think the intention of this is to also check child pages for authors and percolate that up to the publication page, is that correct?

In which case, cool, but I'd like to double check that that is a requirement in spec. I got the impression from talks that some reports will define authors at the 'root' level, and some at the 'article' level, and so it would be up to editors in the CMS to define where they want author stuff to show. 

If not, this is a handy feature to have, and I'd want to make sure we communicate it as available functionality. *party emoji*

KalobTaulien replied:
I think this is a blocker. We need clarification before proceeding with this one.

easherma replied:
After design/dev sync today we confirmed: 

* Authors are optional and can be added to PublicationPage or ArticlePage
* Both pages will show Authors if they are selected for that page
* They will not show authors that have been selected for other pages

easherma commented on network-api/networkapi/wagtailpages/templates/wagtailpages/publication_page.html:
```diff
@@ -32,8 +32,29 @@ <h1 class="h1-heading">
 
-                    {% if page.publication_file %}
 
+                    <div class="blog-authors d-flex flex-wrap align-items-center">
```
Is this template copy-pasted from elsewhere or is it different? I'm wondering if we have the opportunity to make a reusable template block here.

KalobTaulien replied:
It sure is! It came from the homepage work that just got merged into master. The fragment is `blog_authors.html`. I just threw this in here for now and was thinking about refactoring this when we do the official FE work.

easherma replied:
Makes sense to me!

KalobTaulien commented:
@easherma Updated this PR so it uses an Orderable to explicitly select Authors per Publication page. Work is in the latest commit https://github.com/mozilla/foundation.mozilla.org/pull/5135/commits/3dbe975fec0e518ff07d8ea95f0fd81b5264d8d2 — could you quickly review once more, and if you're happy merge in?

easherma commented on network-api/networkapi/wagtailpages/pagemodels/publications/publication.py:
```diff
@@ -72,20 +95,16 @@ class PublicationPage(FoundationMetadataPageMixin, Page):
-        FieldPanel('publication_date'),
-        ImageChooserPanel('hero_image'),
-        DocumentChooserPanel('publication_file'),
+        MultiFieldPanel([
```
minor quibbles:
* if Authors are optional and therefore not always entered, I'd rather have them below other required or commonly entered stuff like the Hero
* curious about the motivation to put the other hero fields into a multifield panel, im assuming thats to group things a bit more semantically. In which case wouldn't authors be apart of that since they are also in the hero?

KalobTaulien replied:
No reason. It's because I copy/pastas'd from some other authors work. Will move `authors` into the hero `MultiFieldPanel`

easherma commented on network-api/networkapi/wagtailpages/pagemodels/publications/publication.py:
```diff
@@ -72,20 +95,16 @@ class PublicationPage(FoundationMetadataPageMixin, Page):
-        ImageChooserPanel('hero_image'),
-        DocumentChooserPanel('publication_file'),
+        MultiFieldPanel([
+            InlinePanel("authors", label="Author", min_num=1)
```
Authors should be totally optional, `min_num=None` would be more appropriate and prevent editors/authors from getting confused about a required field if it isn't required.

easherma commented on network-api/networkapi/wagtailpages/pagemodels/publications/article.py:
```diff
@@ -44,6 +67,9 @@ class ArticlePage(FoundationMetadataPageMixin, Page):
 
     content_panels = Page.content_panels + [
+        MultiFieldPanel([
+            InlinePanel("authors", label="Author", min_num=1)
```
Same here, no min_num

easherma commented on network-api/networkapi/wagtailpages/pagemodels/publications/article.py:
```diff
@@ -23,6 +27,25 @@
+        BlogAuthor,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=False
```
Authors are optional

KalobTaulien replied:
Yep, these are optional. It's set by the `min_num` value with `null=True`. Blank just means if an author Orderable "Add Author" button is clicked, then make sure it's filled out instead of providing an empty value.

easherma commented on network-api/networkapi/wagtailpages/pagemodels/publications/publication.py:
```diff
@@ -1,19 +1,42 @@
+        BlogAuthor,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=False
```
Authors should be optional, according to specs not every publication is going to have specific authors cited

KalobTaulien commented:
@easherma Can you give this a once over again? I added the additional copy field, moved the fields you suggested and removed authors being mandatory. (Just the last 4 commits)
